lavin/mm_adapter.py

```diff
@@ -32,9 +35,14 @@ def __init__(self, in_features=768, hidden_dim=8, groups=2, scale=1, t=10.):
         nn.init.zeros_(self.conv_D.bias)
 
     def forward(self, x, weights=None):
-        with autocast():
+        with autocast(dtype=self.precision):
             if weights is None:
-                weights = torch.softmax(self.expert_weights(x[:, 0]) / self.t, -1).half()
+                weights = torch.softmax(self.expert_weights(x[:, 0]) / self.t, -1)
+                if self.precision == torch.float16:
```

Comment:
if weights are float16, then float16 * bfloat16 -> float32!!!
Therefore, we need to convert weights into bfloat16 if needed.